builder/job.go

```diff
@@ -121,24 +124,52 @@ func Build(d *daemon.Daemon, buildConfig *Config) error {
 	} else if urlutil.IsURL(buildConfig.RemoteURL) {
 		f, err := httputils.Download(buildConfig.RemoteURL)
 		if err != nil {
-			return err
+			return fmt.Errorf("Error downloading remote context %s: %s\n", buildConfig.RemoteURL, err)
 		}
 		defer f.Body.Close()
-		dockerFile, err := ioutil.ReadAll(f.Body)
-		if err != nil {
-			return err
+
+		plen := int(f.ContentLength)
+		if plen > maxPreambleLength {
```

Comment:
If there is no Content-Length value then -1 will be returned on line 131 - I think you need to check for that and take some action based on it. Perhaps set plen to maxPreambleLength?